lib/pos_consensus_controller.go

```diff
@@ -211,3 +211,109 @@ func (cc *ConsensusController) HandleHeader(pp *Peer, msg *MsgDeSoHeader) {
+	}
+
+	// Fetch the validator set at the new tip
+	validatorSetAtTipBlock, err := utxoViewAtTipBlock.GetFullSnapshotValidatorSet()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching validator set for tip block: %v", err)
+	}
+
+	// Bundle the tip block and validator set together
+	tipBlockWithValidators := consensus.BlockWithValidatorList{
+		Block:         tipBlockHeader,
+		ValidatorList: ValidatorEntriesToConsensusInterface(validatorSetAtTipBlock),
+	}
+
+	// Fetch the remaining safe blocks that are eligible to be extended from by the next incoming tip block
+	safeBlocks, err := cc.blockchain.getSafeBlocks()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching safe blocks: %v", err)
+	}
+
+	// Fetch the validator sets for the safe blocks
+	safeBlocksWithValidators := make([]consensus.BlockWithValidatorList, len(safeBlocks))
+	for ii, safeBlock := range safeBlocks {
+		safeBlockHash, err := safeBlock.Hash()
+		if err != nil {
+			return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error hashing safe block: %v", err)
+		}
+
+		utxoViewAtSafeBlock, err := cc.blockchain.getUtxoViewAtBlockHash(*safeBlockHash)
+		if err != nil {
+			return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching UtxoView for safe block: %v", err)
+		}
+
+		validatorSetAtSafeBlock, err := utxoViewAtSafeBlock.GetFullSnapshotValidatorSet()
+		if err != nil {
+			return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching validator set for safe block: %v", err)
+		}
+
+		safeBlocksWithValidators[ii] = consensus.BlockWithValidatorList{
+			Block:         safeBlock,
+			ValidatorList: ValidatorEntriesToConsensusInterface(validatorSetAtSafeBlock),
+		}
+	}
+
+	// Pass the new tip and safe blocks to the FastHotStuffEventLoop
+	if err = cc.fastHotStuffEventLoop.ProcessTipBlock(tipBlockWithValidators, safeBlocksWithValidators); err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error processing tip block locally: %v", err)
+	}
+
+	// Happy path. The block was processed successfully and applied as the new tip.
+	// There are no missing blocks and no error.
+	return nil, nil
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// TODO: delete all of the functions below. They are dummy stubbed out functions
+// needed by ConsensusController, but are implemented in other feature branches.
+// We stub them out here to unblock consensus work.
+////////////////////////////////////////////////////////////////////////////////
+
+func (bc *Blockchain) getUtxoViewAtBlockHash(blockHash BlockHash) (*UtxoView, error) {
+	return nil, errors.New("getUtxoViewAtBlockHash: replace me with a real implementation later")
+}
+
+func (bc *Blockchain) getSafeBlocks() ([]*MsgDeSoHeader, error) {
+	return nil, errors.New("getSafeBlocks: replace me with a real implementation later")
+}
```

Comment:
ah yes, if you want the names we have in `feature/pos-process-block`, you can use the following signatures:

```
func (bc *Blockchain) GetSafeBlocks() ([]*BlockNode, error) {
```
Looks like you have the correct signature for `getUtxoViewAtBlockHash`

Reply:
Discussed offline. We'll keep the function signature for this as `func (bc *Blockchain) getSafeBlocks() ([]*MsgDeSoHeader, error)` so that the Blockchain struct doesn't expose BlockNodes where they're not necessary